### PR TITLE
Add fine-grained RBACs for gardenlet in the seed cluster

### PR DESCRIPTION
**Pull request by vpnachev**
**How to categorize this PR?**

/area security
/kind enhancement

**What this PR does / why we need it**:
Configures `gardenlet` with more fine-grained permissions in the seed cluster.

**Which issue(s) this PR fixes**:
Part of #1724 

**Special notes for your reviewer**:

**Release note**:

```other operator
The RBACs for the `gardenlet` in the seed cluster are now fine-grained defined and limited as much as possible.
```

### CODE REVIEW

**Review by rfranzke — COMMENTED** [review]
/assign
/invite @stoyanr

**Review by rfranzke — COMMENTED** [review]
Can you prefix all roles/rolebindings with `gardener.cloud:system:gardenlet:`?

**Comment by vpnachev** [comment]
> Can you prefix all roles/rolebindings with `gardener.cloud:system:gardenlet:`?

Done.

File: charts/gardener/gardenlet/charts/runtime/templates/rbac.yaml [hunk 1]
```diff
@@ -1,6 +1,59 @@
 {{- if .Values.global.gardenlet.enabled }}
 # ClusterRole defines the required permissions for the gardenlet in the seed
-# For now, we provide all privileges for simplicity reasons.
+# For now, we provide escalation privileges required by gardener resoruce manager.
```
Comment (ialidzhikov) [inline charts/gardener/gardenlet/charts/runtime/templates/rbac.yaml]: This file now contains 12 resources. According to the helm best practices (ref https://helm.sh/docs/chart_best_practices/templates/):

> Each resource definition should be in its own template file.

Does it make sense to split this huge template file into dedicated template files. WDYT?
I find it hard to follow when everything is in single template file 😅 .

File: charts/gardener/gardenlet/charts/runtime/templates/rbac.yaml [hunk 2]
```diff
@@ -15,17 +68,327 @@ metadata:
...
+  - create
+- apiGroups:
+  - admissionregistration.k8s.io
+  resources:
+  - mutatingwebhookconfigurations
+  resourceNames:
+  - vpa-webhook-config-seed
+  verbs:
+  - get
+  - update
```
Comment (ialidzhikov) [inline charts/gardener/gardenlet/charts/runtime/templates/rbac.yaml]: When the Seed is with

```yaml
spec:
  settings:
    verticalPodAutoscaler:
      enabled: false
```

then gardenlet will try to delete this webhook. 

https://github.com/gardener/gardener/blob/9df08ae3ae865f96db4161eb4ad01ced73107f3e/pkg/operation/common/utils.go#L280.

```
time="2021-03-26T13:56:05+02:00" level=error msg="seed is not yet ready: condition \"Bootstrapped\" has invalid status False (expected True) due to BootstrappingFailed: mutatingwebhookconfigurations.admissionregistration.k8s.io \"vpa-webhook-config-seed\" is forbidden: User \"system:serviceaccount:garden:gardenlet\" cannot delete resource \"mutatingwebhookconfigurations\" in API group \"admissionregistration.k8s.io\" at the cluster scope" operation=reconcile shoot=garden-foo/aws-local
```

```suggestion
  - vpa-webhook-config-seed
  verbs:
  - get
  - update
  - delete
```

File: pkg/gardenlet/controller/managedseed/managedseed_actuator.go [hunk 3]
```diff
@@ -165,6 +165,31 @@ func (a *actuator) createOrUpdateGardenNamespace(ctx context.Context, shootClien
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name: namespaceName,
+		},
+	}
+
+	err := shootClient.Client().Get(ctx, kutil.Key(sniIngressNamespace.Name), sniIngressNamespace)
+	if apierrors.IsNotFound(err) {
+		return shootClient.Client().Create(ctx, sniIngressNamespace)
+	}
+	return err
```
Comment (ialidzhikov) [inline pkg/gardenlet/controller/managedseed/managedseed_actuator.go:178]:
```suggestion
	_, err := controllerutil.CreateOrUpdate(ctx, shootClient.Client(), sniIngressNamespace, func() error {
		return nil
	})

	return err
```

**Review by ialidzhikov — COMMENTED** [review]
Thank you for your hard work on this topic! Appreciate it!
It seems that I will need more time than I initially expected to setup a proper env to try out this change.
I am leaving few comments inline for now.

File: charts/gardener/gardenlet/charts/runtime/templates/rbac.yaml [hunk 4]
```diff
@@ -15,17 +68,327 @@ metadata:
...
+  - customresourcedefinitions
+  resourceNames:
+  - hvpas.autoscaling.k8s.io
+  verbs:
+  - delete
+- apiGroups:
+  - apps
+  resources:
+  - deployments
+  - statefulsets
```
Comment (ialidzhikov) [inline charts/gardener/gardenlet/charts/runtime/templates/rbac.yaml]: gardenlet should be able to list replicasets and get kube-apiserver Pod logs. Ref https://github.com/gardener/gardener/pull/3353.

```
time="2021-03-26T16:46:08+02:00" level=error msg=Error error="failure to find the newest pod for deployment to read the logs: replicasets.apps is forbidden: User \"system:serviceaccount:garden:gardenlet\" cannot list resource \"replicasets\" in API group \"apps\" in the namespace \"shoot--foo--bar\": retry failed with context deadline exceeded, last error: kube-apiserver deployment has outdated replicas" flow="Shoot cluster reconciliation" operation=reconcile shoot=garden-foo/bar task="Waiting until Kubernetes API server reports readiness"
```

You can reproduce this by specifying a wrong feature gate for the kube-apiserver.

File: charts/gardener/gardenlet/charts/runtime/templates/rbac.yaml [hunk 5]
```diff
@@ -15,17 +68,327 @@ metadata:
...
 rules:
 - apiGroups:
-    - "*"
+  - ""
+  resources:
+  - pods
+  verbs:
+  - list
+  - watch
+  - delete
```
Comment (ialidzhikov) [inline charts/gardener/gardenlet/charts/runtime/templates/rbac.yaml]: gardenlet needs deletecollection for pods. 

https://github.com/gardener/gardener/blob/9df08ae3ae865f96db4161eb4ad01ced73107f3e/pkg/operation/botanist/controlplane.go#L1015-L1023

```
time="2021-03-26T17:17:37+02:00" level=error msg=Error error="pods is forbidden: User \"system:serviceaccount:garden:gardenlet\" cannot deletecollection resource \"pods\" in API group \"\" in the namespace \"shoot--foo--bar\"" flow="Shoot cluster reconciliation" operation=reconcile shoot=garden-foo/bar task="Restart control plane pods"
```

File: pkg/gardenlet/controller/managedseed/managedseed_actuator.go [hunk 6]
```diff
@@ -288,22 +313,23 @@ func (a *actuator) deployGardenlet(ctx context.Context, shootClient kubernetes.I
...
-	// Ensure seed secrets are deleted
 	if err := a.ensureSeedSecretsDeleted(ctx, &gardenletConfig.SeedConfig.SeedTemplate.Spec, managedSeed); err != nil {
 		return err
 	}
 
+	if err := a.deleteSNIIngressNamespace(ctx, shootClient, *gardenletConfig.SNI.Ingress.Namespace); err != nil {
+		return fmt.Errorf("failed to delete SNI ingress namespace %s in shoot %s: %w", *gardenletConfig.SNI.Ingress.Namespace, kutil.ObjectName(shoot), err)
+	}
+
 	// Prepare gardenlet chart values
```
Comment (stoyanr) [inline pkg/gardenlet/controller/managedseed/managedseed_actuator.go]: Shouldn't we check `gardenletConfig.SNI.Ingress.Namespace` is non-nil before dereferencing it (see above)?

What is this namespace supposed to contain? Typically namespace deletion may take some time, e.g. deletion of the `garden` namespace takes sometimes > 5 minutes. Should we ensure it's deleted before the managed seed itself is deleted, like for the `garden` namespace?

Note: I plan to refactor the deletion code in the near future to get rid of the current requeue-through-error behavior. So I could also take care to wait for this namespace if you would rather not do it in this PR, just let me know if you think it's a good idea.

File: pkg/gardenlet/controller/managedseed/managedseed_actuator.go [hunk 7]
```diff
@@ -257,23 +282,23 @@ func (a *actuator) ensureSeedDeleted(ctx context.Context, managedSeed *seedmanag
...
 		return err
 	}
 
-	// Create or update seed secrets
 	if err := a.createOrUpdateSeedSecrets(ctx, &gardenletConfig.SeedConfig.SeedTemplate.Spec, managedSeed, shoot); err != nil {
 		return err
 	}
 
-	// Prepare gardenlet chart values
+	if err := a.createSNIIngressNamespace(ctx, shootClient, *gardenletConfig.SNI.Ingress.Namespace); err != nil {
```
Comment (stoyanr) [inline pkg/gardenlet/controller/managedseed/managedseed_actuator.go]: Shouldn't we check `gardenletConfig.SNI.Ingress.Namespace` is non-nil before dereferencing it? Actually, `SNI` and `Ingress` are also pointers and should be checked as well I believe. Perhaps you could introduce a helper function to properly get this namespace?

It's true that the resource has defaults to always set this. However, these defaults are only applied when an actual resource is created, but not when the resource is deserialized as part of a template - like in this case. So all of these pointers could well be nil I believe.

File: pkg/gardenlet/controller/managedseed/managedseed_actuator.go (hunk 3 above)
Comment (stoyanr) [reply pkg/gardenlet/controller/managedseed/managedseed_actuator.go:178]: I also think what @ialidzhikov proposes should be better.

**Review by stoyanr — CHANGES_REQUESTED** [review]

File: pkg/gardenlet/controller/managedseed/managedseed_actuator.go (hunk 3 above)
Comment (vpnachev) [reply pkg/gardenlet/controller/managedseed/managedseed_actuator.go:178]: This is done on purpose. In some cases the Istio installation is managed by other component, not by gardener and I do not want to touch the namespace at all. This is why this create call must be invoked only once only if the namesapce does not exist.

**Comment by ialidzhikov** [comment]
Thanks for updating the PR. I will soon go over it once again. Sorry for the delay.

File: charts/gardener/gardenlet/charts/runtime/templates/clusterrole-gardenlet.yaml [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+  - delete
+  - get
+  - patch
+  - update
+- nonResourceURLs:
+  - /healthz
+  - /version
+  verbs:
+  - get
+{{- end }}
```
Comment (ialidzhikov) [inline charts/gardener/gardenlet/charts/runtime/templates/clusterrole-gardenlet.yaml]: You could add an empty line to the end of the file.

File: charts/gardener/gardenlet/charts/runtime/templates/rolebinding-default-gardenlet.yaml [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    heritage: "{{ .Release.Service }}"
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: gardener.cloud:system:gardenlet
+subjects:
+- kind: ServiceAccount
+  name: "{{ required ".Values.global.gardenlet.serviceAccountName is required" .Values.global.gardenlet.serviceAccountName }}"
+  namespace: garden
+{{- end }}
```
Comment (ialidzhikov) [inline charts/gardener/gardenlet/charts/runtime/templates/rolebinding-default-gardenlet.yaml]: You could add an empty line to the end of the file.

File: charts/gardener/gardenlet/charts/runtime/templates/rolebinding-garden-gardenlet.yaml [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+    heritage: "{{ .Release.Service }}"
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: gardener.cloud:system:gardenlet
+subjects:
+- kind: ServiceAccount
+  name: "{{ required ".Values.global.gardenlet.serviceAccountName is required" .Values.global.gardenlet.serviceAccountName }}"
+  namespace: garden
+{{- end }}
```
Comment (ialidzhikov) [inline charts/gardener/gardenlet/charts/runtime/templates/rolebinding-garden-gardenlet.yaml]: You could add an empty line to the end of the file.

File: charts/gardener/gardenlet/charts/runtime/templates/clusterrole-gardenlet.yaml [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+metadata:
+  name: gardener.cloud:system:gardenlet
+  labels:
+    gardener.cloud/role: gardenlet
+    app: gardener
+    role: gardenlet
+    chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+    release: "{{ .Release.Name }}"
+    heritage: "{{ .Release.Service }}"
+rules:
```
Comment (ialidzhikov) [inline charts/gardener/gardenlet/charts/runtime/templates/clusterrole-gardenlet.yaml:16]: https://github.com/gardener/gardener/pull/3811 was recently merged and it lists all Seed nodes to decide on seed-admission-controller Deployment replicas.

```suggestion
rules:
- apiGroups:
  - ""
  resources:
  - nodes
  verbs:
  - list
  - watch
```

```
time="2021-04-09T13:16:29+03:00" level=error msg="seed is not yet ready: condition \"Bootstrapped\" has invalid status False (expected True) due to BootstrappingFailed: 1 error occurred:\n\t* task \"Deploying gardener-seed-admission-controller\" failed: nodes is forbidden: User \"system:serviceaccount:garden:gardenlet\" cannot list resource \"nodes\" in API group \"\" at the cluster scope\n\n" operation=reconcile shoot=garden-foo/aws-local
```

File: charts/gardener/gardenlet/charts/runtime/templates/clusterrole-gardenlet.yaml [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+  resourceNames:
+  - hvpas.autoscaling.k8s.io
+  verbs:
+  - delete
+- apiGroups:
+  - apps
+  resources:
+  - deployments
+  - statefulsets
+  - replicasets
```
Comment (ialidzhikov) [inline charts/gardener/gardenlet/charts/runtime/templates/clusterrole-gardenlet.yaml:117]: The case described in https://github.com/gardener/gardener/pull/3763#discussion_r602357037 still fails:

```
time="2021-04-09T13:37:03+03:00" level=error msg="Failed to reconcile Shoot \"aws-local\": flow \"Shoot cluster reconciliation\" encountered task errors: [task \"Waiting until Kubernetes API server reports readiness\" failed: failure to read the logs: pods \"kube-apiserver-7968db85fd-gtd45\" is forbidden: User \"system:serviceaccount:garden:gardenlet\" cannot get resource \"pods/log\" in API group \"\" in the namespace \"shoot--foo--aws-local\": retry failed with context deadline exceeded, last error: kube-apiserver deployment has outdated replicas]" operation=reconcile shoot=garden-foo/aws-local
```

File: charts/gardener/gardenlet/charts/runtime/templates/clusterrole-gardenlet.yaml [hunk 13]
```diff
@@ -0,0 +1,293 @@
+{{- if .Values.global.gardenlet.enabled }}
```
Comment (ialidzhikov) [inline charts/gardener/gardenlet/charts/runtime/templates/clusterrole-gardenlet.yaml:1]: With https://github.com/gardener/gardener/pull/3535 the default resource lock was changed to leases. Currently I don't see any rule for leases. I think we should add a rule like:

```yaml
- apiGroups:
  - coordination.k8s.io
  resources:
  - leases
  verbs:
  - create
- apiGroups:
  - coordination.k8s.io
  resources:
  - leases
  resourceNames:
  - gardenlet-leader-election
  verbs:
  - get
  - watch
  - update
```

**Comment by vpnachev** [comment]
/status author-action

I found some more things that need to be done, but will not be able to work on it next.
Will update here when it is ready for next review round.

**Review by danielfoehrKn — COMMENTED** [review]
I think this related [part of the documentation](https://github.com/gardener/gardener/blob/69efab1bd8a415cb4b280699e6bd62149a7ecd67/docs/deployment/deploy_gardenlet_manually.md#kubeconfig-for-the-seed-cluster) needs an update as no longer admin privileges are required in the Seed cluster.

File: charts/gardener/gardenlet/charts/runtime/templates/clusterrole-gardenlet.yaml [hunk 14]
```diff
@@ -0,0 +1,306 @@
...
+  - watch
+- apiGroups:
+  - autoscaling.k8s.io
+  resources:
+  - hvpas
+  resourceNames:
+  - etcd-events
+  - etcd-main
+  - kube-apiserver
+  - loki
```
Comment (ialidzhikov) [inline charts/gardener/gardenlet/charts/runtime/templates/clusterrole-gardenlet.yaml:154]: FYI, https://github.com/gardener/gardener/pull/3903 (if merged before this PR) would require extending this list:

```suggestion
  resourceNames:
  - etcd-events
  - etcd-main
  - kube-apiserver
  - loki
  - aggregate-prometheus
  - prometheus
```

**Comment by rfranzke** [comment]
What is the status of this PR?

**Comment by vpnachev** [comment]
I am out of capacity last days, and probably will be so in the next 2-3 weeks.
Therefore I am closing it for now and will re-open once I address the comments.
